### PR TITLE
Set up Live Preview functionality

### PR DESCRIPTION
**Pull request by j0nas**
# Set up Live Preview functionality

## Intent

- Implement support for [Live Preview](https://www.sanity.io/guides/nextjs-live-preview) 
- Fix [this Shortcut Story](https://app.shortcut.com/sanity-io/story/14246/set-up-preview-in-studio)

## Description

This PR implements Live Preview support for [Route](https://structured-content-2022-studio-1plwyyvga.sanity.build/desk/route) entities in Studio by following the linked guide. Assuming the general setup in this PR is OK, a follow-up PR can expand `resolveProductionUrl` to also work for `page`-type documents, but this would require a bit of GROQ cleverness which is out of scope for this PR.

Note the changed placement of the `...,` operator in the `QUERY` in _web/pages/[[...slug]].tsx_; I spent a good while trying to figure out why the client-side query errored out before realizing that the server- and client-side Sanity clients interpret the results of such a positioned `...` operator differently. Moving this operator to the top fixed the issue. (cc @kmelve, maybe fixing this upstream would save others some potential pain.)

## Testing this PR

1. Add a _.env.development_ file to the root of your repo (_.env_ files are not picked up!)
2. Enter `SANITY_STUDIO_PREVIEW_SECRET='<VALUE>'` into the file, replacing `<VALUE>` with the value of the [environment variable set up in Vercel](https://vercel.com/sanity-io/structured-content-2022-studio/settings/environment-variables).
3. Start the _/studio_ and _/web_ projects separately (`cd /[web/studio] && yarn && yarn dev`)
4. Navigate to e.g. [/shared-sections-test](http://localhost:3000/shared-sections-test) and verify that you get additional draft content (e.g., I added a "This is a test page" Hero heading to the [Studio document](https://structured-content-2022-studio-1plwyyvga.sanity.build/desk/page;345118ad-3b89-49c9-9ef1-c98f1db78cce)) loaded a couple seconds after the initial load
5. Open Studio locally and navigate to e.g. the [/shared-section-test](http://localhost:3333/desk/route;36c8bda6-e3a3-4671-a260-cdaf3dcd2e43) document
6. Click the top-right meatballs menu and verify that a "Open preview" option is present
7. Click said option, and verify that you're taken to http://localhost:3000/shared-sections-test, which again should load draft content post-load
8. Change some content in the document and verify that the content of the page you opened in your browser changes too

("Open preview" does not redirect to correct URLs for Preview Deploys, instead redirecting to the staging API (where this functionality is not yet available. We also need to remember to update this URL post-launch to point to production.)

### CODE REVIEW

**Review by oyvind-stenhaug — APPROVED** [review]
This is not functionality that I have any prior familiarity with, so maybe Knut can review it? If not then I skimmed through the diff and can "rubber-stamp" it if it works for you.

Here I'm just getting "Invalid secret token", but maybe it's something wrong on my side. Personally I don't really need or want live preview anyway, so I don't mind.

**Comment by kmelve** [comment]
Looks great - just some smol nitpicks. I wonder if the GROQ behavior can have been tied API versions. I noticed that we haven't specified the API version for the website client, while `groq-store` (which powers the preview) runs on a >V1 version.  I've pushed a commit for it to this branch.

File: web/pages/[[...slug]].tsx [hunk 1]
```diff
@@ -237,14 +245,17 @@ const Route = ({
   );
 };
 
-export async function getServerSideProps({ params }) {
+export const getServerSideProps: GetServerSideProps = async ({
```
[the remark this answers is not in the document]
Comment (j0nas) [reply web/pages/[[...slug]].tsx:248]: Converting to using getStaticProps requires also setting up webhooks to rebuild pages on Sanity content updates, and is out of scope for this PR. That bit is covered by [this](https://app.shortcut.com/sanity-io/story/14538/rebuild-page-when-respective-sanity-content-is-updated) Shortcut Story.

File: studio/README.md [hunk 2]
```diff
@@ -7,3 +7,14 @@ Now you can do the following things:
...
 - [Join the community Slack](https://slack.sanity.io/?utm_source=readme)
 - [Extend and build plugins](https://www.sanity.io/docs/content-studio/extending?utm_source=readme)
+
+## Enabling Live Preview
+
+[Live Preview][0] is a great way to test your content in real-time. Your local page can reflect your Studio drafts.
+
+You can enable Live Preview by adding a _.env.development_ file to the _/studio_ directory with a `SANITY_STUDIO_PREVIEW_SECRET` variable.
+Fetch the value for this key [in the "Environment Variables" section of the Vercel project][1]. Once added, restart the project locally.
+Now, pages will load draft content client-side after the initial render.
```
[the remark this answers is not in the document]
Comment (j0nas) [reply studio/README.md]: Fixed in e5e2344).

File: web/pages/[[...slug]].tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (kmelve) [reply web/pages/[[...slug]].tsx:248]: Agree that it's out of scope for this PR. I'll follow up in the story

File: studio/src/resolveProductionUrl.ts [hunk 3]
```diff
@@ -0,0 +1,17 @@
+const remoteUrl = 'https://structured-content-2022.sanity.build/'
+const localUrl = 'http://localhost:3000'
+
+export default function resolveProductionUrl(doc) {
```
[the remark this answers is not in the document]
Comment (j0nas) [reply studio/src/resolveProductionUrl.ts:3]: Great suggestion! How about a806c50?

**Review by kmelve — COMMENTED** [review]
LGTM!

**Review by kmelve — APPROVED** [review]

**Pull request merged**